src/fsharp/TypeChecker.fs

```diff
@@ -265,7 +265,10 @@ type TcEnv =
       eInternalsVisibleCompPaths: CompilationPath list // internals under these should be accessible
 
       /// Mutable accumulator for the current module type 
-      eModuleOrNamespaceTypeAccumulator: ModuleOrNamespaceType ref 
+      eModuleOrNamespaceTypeAccumulator: ModuleOrNamespaceType ref
+
+      /// Error Message for type errors
+      specializedErrorMessage : ((string*string) -> string) option
```

Comment:
Extending TcEnv is a bit problematic. for two reasons

* We really need to be reducing the size of TcEnv instead of increasing it.  One technique would be to split it between hot (varying fields) and cold (usually empty/None/...) fields, and only allocate the cold block when one of the fields gets a non-default value.

* The entry you've added needs to be reset to "None" at an appropriate point.  For example, when you go into a lambda, or perhaps any expression that's not a control construct i.e. not an if/then/else, match, try/with, try/finally.

Adding a function also feels a bit wrong - I'd rather have a more explicit representation of the information being added here, so we know more clearly when the information no longer applies (e.g. in the cases mentioned above), or we can work out how it interacts with other "expression context" information we are bound to add in the future.  So it feels right to change this to be a "eContextInfo" field that contains a union between various expression contexts, with a clear definition in the ``///`` comments for the cases as to what it means to be in that context.  The two contexts added for this PR would presumably be ``ElseBranch`` and ``IfBranchWithoutElse``